_posts/2021-12-08-kafkasql-storage-and-security.md

```diff
@@ -0,0 +1,471 @@
+---
+layout: post
+title: "Resolving a bug in KafkaSQL storage for Apicurio Registry"
+date:   2021-12-08 12:00:00
+author: jsenko
+categories: blog registry security
+---
+
+One of the major changes introduced in Apicurio Registry `2.x` were improvements in the supported storage options.
+We have kept the possibility of using both PostgreSQL database and Kafka to store your data, but the internals of the 
+storage options have been improved, and in some ways simplified.
+
+We have received great feedback from the users of the new KafkaSQL storage option, but recently several users have 
+reported an interesting bug when using KafkaSQL. We have investigated the problem, found the cause, and would
+like to take this opportunity to share the details with the Apicurio community.
+
+In this blog post, we will take a look at the KafkaSQL storage option in more detail. 
+We'll describe the bug and the solution, and provide a temporary workaround for users 
+until they are able to upgrade to the patched version `2.1.4.Final`.
+
+---
+
+# KafkaSQL Storage
+
+A major use-case for Apicurio Registry is to manage and distribute schemas for messages in Kafka topics. 
+This means that a lot of our users have a Kafka cluster available, so supporting it as a storage option has been a priority for us.
+
+KafkaSQL storage shares common logic with the standard SQL storage option, 
+but does not use a central SQL database to store Apicurio Registry data.
+Instead, each node uses a local SQL database. To keep the replicas in sync, each change in one of the
+local databases is replicated to the other nodes in the cluster using a Kafka topic.
+This makes the read operations on the data straightforward, and allows us to keep the commonality with the SQL persistence option.
+
+The design of the Kafka messages closely follows the tables that are used both in the local SQL database, 
+and the database for the standard SQL storage.
+
+Each Kafka message has a key and a corresponding value. Kafka itself is agnostic about their content, and treats them as a byte array.
+In the KafkaSQL storage, they are represented by objects that are serialized into a JSON String, 
+prepended with marker bytes used to recognize message type. 
+
+In general, each message value class has a corresponding key class. 
+Classes for the keys are mainly responsible for determining the partition that the message is going to reside in, 
+the tenant information in case of a multi-tenant deployment, and identifying information for the object that the message value represents. 
```

Comment:
I was expecting something about uniqueness and log compaction here.  Perhaps that's mentioned specifically later.  If so I'll come back and delete this comment. :)

Reply:
I didn't find anything, so I think somewhere we need some explanation about how the Key is also vitally important for log compaction.  If two messages have the same key, then only the most recent message with that key will be retained when the kafka topic has log compaction enabled.  So it's OK to remove some messages from the topic as long as doing so will not impact the overall replication of state in the DB.  Thus we must be very careful and thoughtful about the uniqueness of our message keys **or** we cannot reliably use log compaction.

Reply:
I think I mentioned it in the section describing how compaction works

Reply:
It's a bit later, but may need to be mentioned earlier

Reply:
Ah - now I see the section on Log Compaction... :)